### ISSUE TITLE
Copy should copy symlinks, not resolve them

### ISSUE THREAD

**Issue by novaterata** [gradle/gradle#3982]
By always resolving symlinks, final artifacts are ballooning in size

### Expected Behavior
Preserve symlinks, don't duplicate files

### Current Behavior
Files are duplicated

### Context
Because the Copy task (and I presume the Zip and Tar tasks) resolves symlinks instead of preserving them, my 3rd party dependency (that I have no control over) goes from 70MB to 200MB as they have many .so files with version-less or major-version-only aliases using symlinks. That's 130MB I'd prefer to not have in my Docker image on Azure. I'm going to have replace all my Copy and Tar tasks with platform specific Exec tasks now.

### Steps to Reproduce (for bugs)
Create a directory with a large number of symlinks
Copy that directory in a Copy task

### Your Environment
Ubuntu 17.10 64-bit
Gradle 4.3.1

**Comment by novaterata** [comment]
tarTree also extracts symlinks as 0 length files

**Comment by novaterata** [comment]
https://stackoverflow.com/questions/11866597/how-to-prevent-cmake-from-creating-a-symlink-for-shared-libraries-on-install  For reference the symlinks I have seem to be some common CMake behavior

**Comment by hugares** [comment]
Until this issue is resolved, here is a workaround for linux users. This custom task is an extension of Gradle copy task which preserves the symlinks:

```
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

import org.gradle.api.tasks.Copy

class CopyWithSymlink extends Copy {
    public CopyWithSymlink() {
        super();
        eachFile { details ->
            Path sourcePath = FileSystems.getDefault().getPath(details.file.path)
            if(Files.isSymbolicLink(sourcePath)) {
                details.exclude()
                Path destinationPath = Paths.get("${destinationDir}/${details.relativePath}")
                if(Files.notExists(destinationPath.parent)) {
                    mkdir destinationPath.parent
                }
                project.exec {
                    commandLine 'ln', '-sf', Files.readSymbolicLink(sourcePath), destinationPath
                }
            }
        }
    }
}
```

**Comment by magneticflux-** [comment]
I believe to resolve this issue, the internal `Copy` actions have to be changed from using `InputStream`s and `OutputStream`s to using NIO filesystem-aware `Path`s. This may be possible in a backwards-compatible manner by overloading some methods like `FileTreeElement::copyTo(OutputStream)` to take `Path`s, and then using the `Path` method whenever possible.

**Comment by martinda** [comment]
Some similarity with the unix tar command would be useful in solving this issue:
* do not fail when the link cannot be resolved, just copy the symbolic link itself even if it does not dereference
* have an option to follow symbolic links
* have an option to follow the hard links

Options may be necessary to preserve the old behavior (no matter how bad) so as to minimize incompatibility.

**Comment by dchesterman** [comment]
Is this going to fix the issue with tartree's and the like during copy as well? Specifically with archived symlinks being empty when expanded?

For context, if you set up something like the following 
```
mkdir -p archiveFolder
echo "Hello World" >> archiveFolder/realFile.txt
ln -s realFile.txt archiveFolder/linkFile.txt
cat archiveFolder/linkFile.txt
tar -czvf archive.tgz archiveFolder
```
and have a build.gradle file with 
```
task linkTestExpand(type: Copy) {
    from(tarTree(project.file('archive.tgz')))
    into project.file('expand')
}
```
the link file is empty
```
cat expand/archiveFolder/linkFile.txt
```
I think this is a similar reported issue https://discuss.gradle.org/t/expanding-a-tar-gz-with-tartree-loses-symlinks/30285

**Comment by pedrolamarao** [comment]
I have just hit this problem while trying to expand the git-for-windows bzip2 distribution. The tar tree contains hard links which Gradle expands into empty files.

**Comment by sschuberth** [comment]
It's a pity that a fix for this hasn't made it into Gradle 8.0, as this is one of the last bits to get provisioning of JDK toolchains fully working.

**Comment by martinda** [comment]
In my case, we are currently trying an local modification to the  [`DefaultDirectoryWalker`](https://github.com/gradle/gradle/blob/master/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java#L62) and building our own Gradle distro to try things out. By changing the `FileVisitOption.FOLLOW_LINKS` to a newly created `FileVisitOption.DONT_FOLLOW_LINKS`, and by changing the behavior as it progresses through the code, we're hoping to get to the bottom of it. I don't quite remember if the `DefaultDirectoryWalker` is called when the `AbstractCopyTask` gets the files from the copySpec or when it zips the files in our custom extension of `AbstractArchiveTask` which we modelled after [SimZip.groovy](https://github.com/paleozogt/symzip-plugin/blob/master/plugin/src/main/groovy/org/paleozogt/gradle/zip/SymZip.groovy), but we see that the `DefaultDirectoryWalker` is where it fails.

We wanted to start somewhere, I hope we're starting in the right place. I would like to contribute a PR if this is successful. Any advice?

**Comment by sschuberth** [comment]
> I would like to contribute a PR if this is successful. Any advice?

You should probably align with @FrauBoes who was recently assigned to https://github.com/gradle/gradle/issues/10676.

**Comment by ysb33r** [comment]
I think the behaviour should be able to turn symbolic link behaviour on or off within any `CopySpec`. This allows the best flexibility. Sometime you want the actual symbolic link to be copied and sometime the whole tree or file it is pointing to.

**Comment by mkrussel77** [comment]
> Until this issue is resolved, here is a workaround for linux users. This custom task is an extension of Gradle copy task which preserves the symlinks:
> 
> ```
> import java.nio.file.FileSystems
> import java.nio.file.Files
> import java.nio.file.Path
> import java.nio.file.Paths
> 
> import org.gradle.api.tasks.Copy
> 
> class CopyWithSymlink extends Copy {
>     public CopyWithSymlink() {
>         super();
>         eachFile { details ->
>             Path sourcePath = FileSystems.getDefault().getPath(details.file.path)
>             if(Files.isSymbolicLink(sourcePath)) {
>                 details.exclude()
>                 Path destinationPath = Paths.get("${destinationDir}/${details.relativePath}")
>                 if(Files.notExists(destinationPath.parent)) {
>                     project.mkdir destinationPath.parent
>                 }
>                 project.exec {
>                     commandLine 'ln', '-sf', Files.readSymbolicLink(sourcePath), destinationPath
>                 }
>             }
>         }
>     }
> }
> ```

The `Files.isSymbolicLink` is returning false when the file is in a zip.  Any recommendation?

**Comment by ysb33r** [comment]
I think the behaviour should be able to turn symbolic link behaviour on or off within any `CopySpec`. This allows the best flexibility. Sometime you want the actual symbolic link to be copied and sometime the whole tree or file it is pointing to.

**Comment by sschuberth** [comment]
> I think the behaviour should be able to turn symbolic link behaviour on or off within any `CopySpec`.

Agreed, introducing a `preserveSymlinks` property to `CopySpec` is also what was suggested by @big-guy over [here](https://github.com/gradle/gradle/pull/12795#issuecomment-621631469). Unfortunately that PR was closed unmerged.

**Comment by ov7a** [comment]
@leonard84 @ysb33r 
Links can be relative or absolute.
I think Gradle can just copy links in both cases. If you want to target the copied directory, then you should use a relative link. If you want to preserve the original target, you should use an absolute link.

**Comment by niloc132** [comment]
According to the final comment on https://github.com/gradle/gradle/pull/25264 before it was closed without merging, the Copy task is going to be redesigned - I'm not specifically hoping for a timetable or milestone, but perhaps an issue to watch to see how this might proceed?

**Comment by marcphilipp** [comment]
This currently breaks auto-provisioned GraalVM JDKs because `native-image` is a symlink that does not work after Gradle extracted the archive.

**Comment by sschuberth** [comment]
> This currently breaks auto-provisioned GraalVM JDKs

True, also see https://github.com/gradle/gradle/issues/28583.
